### DISCUSSION TITLE
EvalEx 3.0 early preview

### DISCUSSION THREAD

**Discussion by uklimaschewski** [ezylang/EvalEx#291] [Announcements]
Hello,

I have just published the source code for the upcoming version 3 of EvalEx.

For version 3, I will move the project to a new repository location: https://github.com/ezylang/evalex-core

Version 3 is not release ready by now, but I think the main work has been done and only some bells and whistles are missing for the release.
At this point in the project, I therefore would like to share the code with you. Hopefully to get some feedback.

Version 3 is a complete rewrite from scratch and I took the opportunity of a new major version to:
* Moved to Java 11, which eases a lot of things.
* Improved the code quality and readability.
* Moved from RPN to AST (Abstract Syntax Tree) for internal representation and evaluation.
* Added full string support.
* Added support for arrays and structures, to handle JSON like data structures.
* Changed the License from MIT to Apache 2.0 (which should not make any difference in using EvalEx).

The existing unit tests provide a code coverage of 100%, but this does not mean that there are no bugs. ;-)
I expect some bugs, especially when it comes to more complex and rare use cases. Testers and bug reports are welcome!
Contributions are also very much appreciated.

General discussion about concepts, features etc. should be started at the new location in the [discussion area](https://github.com/ezylang/evalex-core/discussions)
For bugs, please open an [issue](https://github.com/ezylang/evalex-core/issues).

More administrative things like CI/CD and Sonar code analysis etc. will come soon.

Some things that are missing:
* Some more trigonometric functions, I didn't port them all. Also I plan an evalex-bigmath extension project, which will make use of the [big-math](https://github.com/eobermuhlner/big-math) project from Eric Obermühlner.
* Some more string functions.
* Support for array operations (like adding and multiplying arrays, should be easy).
* Some convenience API functions are missing (setting precision, I'd like to rename this as "precision" is constantly confusing people).
* There are already plans for another extension project, evalex-json, which will allow to use e.g. Jackson to give input to evaluation, e.g. by transforming Java objects to JSON and passing them as parameters to an expression.
* Some more fun with the abstract syntax tree, like creating an "execution plan", or evaluation history.

**Comment by RobertZenz** [comment]
I didn't have much time to look at it, but I'm not a friend of annotations. If they are there to allow lambda/method-reference expressions, I'd rather like to see factory methods for expressions than annotations. Or did I misunderstand why you went for annotations?

**Reply by uklimaschewski** [reply]
Hi @RobertZenz,

If you mean the annotations on operator and function definition, they are only there to configure them. They are completely optional and only used by the abstract classes to set the defaults like precedence and associativity for operators and parameter settings (e.g. lazy parameter or var-arg parameter) for functions.

Here and there I use Lombok annotations, got used to them and I find them extremely useful to remove boilerplate code in Java.

**Comment by nikeee** [comment]
Is #156 something to be considered in the rewrite?
The primary issue for us was the heavy constructor of the Expression as well as non-thread-safety.

I've a PoC library that distinguishes the language (the operators, global functions and constants) from a parsed expression (parsed AST, referenced variables and used language) and the expression factory.
The latter can be used to create multiple variations of executions: Interpreting the AST, compiling to a stack machine and compiling to java byte code, which can be chosen depending on the use-case (mostly, AST is the fastest).
The core point is that each mentioned class is immutable, making everything thread-safe. 

Thus, `eval` can be called from mutliple threads on the same instance, reducing the number of instances and language creations:
```java
Language language = new LanguageBuilder()
    .addOperators(CommonOperators.PEMDAS)
    .addFunctions(new Function(
        "successor",
        /* parameters */ 1,
        (es, parameters) -> parameters[0].add(BigDecimal.ONE))
    )
    .build();

Expression exp = language.parseExpression("a + b + successor(b)", new SyntaxTreeInterpreter(language));

Map<String, BigDecimal> scope = Map.of(
    "a", BigDecimal.TEN,
    "b", BigDecimal.ONE
);

EvaluationSettings settings = new EvaluationSettings(MathContext.DECIMAL32, /* stripTrailingZeros */ true, /* requireAllVariablesInScope */ false);

exp.eval(settings, scope);
```

`scope` is just a `Map<String, BigDecimal>` here. If I'm not mistaking, this would be a more complex object for EvalEx3.
